File: src/com/intellij/advancedExpressionFolding/processor/declaration/PsiVariableExt.java
package com.intellij.advancedExpressionFolding.processor.declaration;

import com.intellij.advancedExpressionFolding.expression.VariableDeclarationImpl;
import com.intellij.advancedExpressionFolding.processor.core.BaseExtension;
import com.intellij.advancedExpressionFolding.processor.util.Helper;
import com.intellij.advancedExpressionFolding.settings.AdvancedExpressionFoldingSettings;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiForeachStatement;
import com.intellij.psi.PsiVariable;
import org.jetbrains.annotations.Nullable;

public class PsiVariableExt extends BaseExtension {

    @Nullable
    public static VariableDeclarationImpl getVariableDeclaration(PsiVariable element) {
        AdvancedExpressionFoldingSettings settings = AdvancedExpressionFoldingSettings.getInstance();
        if (shouldCollapseVariableDeclaration(element, settings)) {
            boolean isFinal = Helper.calculateIfFinal(element);
            return new VariableDeclarationImpl(element, TextRange.create(
                    element.getTextRange().getStartOffset(),
                    element.getTypeElement().getTextRange().getEndOffset()),
                    element.getModifierList() != null && isFinal);
        }
        return null;
    }

    private static boolean shouldCollapseVariableDeclaration(
            PsiVariable element, AdvancedExpressionFoldingSettings settings) {
        return settings.getState().getVarExpressionsCollapse()
                && element.getName() != null
                && element.getTypeElement() != null
                && (element.getInitializer() != null || element.getParent() instanceof PsiForeachStatement)
                && element.getTextRange().getStartOffset() < element.getTypeElement().getTextRange().getEndOffset();
    }
}

File: src/com/intellij/advancedExpressionFolding/processor/declaration/PsiVariableExt.kt
package com.intellij.advancedExpressionFolding.processor.declaration

import com.intellij.advancedExpressionFolding.expression.VariableDeclarationImpl
import com.intellij.advancedExpressionFolding.processor.util.Helper
import com.intellij.advancedExpressionFolding.settings.State
import com.intellij.advancedExpressionFolding.settings.state.IExpressionCollapseState
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiForeachStatement
import com.intellij.psi.PsiVariable

object PsiVariableExt : IExpressionCollapseState by State()() {

    fun getVariableDeclaration(element: PsiVariable): VariableDeclarationImpl? {
        if (!shouldCollapseVariableDeclaration(element)) {
            return null
        }
        val isFinal = Helper.calculateIfFinal(element)
        val endOffset = element.typeElement?.textRange?.endOffset ?: return null
        return VariableDeclarationImpl(
            element,
            TextRange.create(element.textRange.startOffset, endOffset),
            element.modifierList != null && isFinal
        )
    }

    private fun shouldCollapseVariableDeclaration(
        element: PsiVariable
    ): Boolean {
        val typeElement = element.typeElement ?: return false
        return varExpressionsCollapse &&
            element.name != null &&
            (element.initializer != null || element.parent is PsiForeachStatement) &&
            element.textRange.startOffset < typeElement.textRange.endOffset
    }
}
